File: back-end/app/models/task.py
"""
任务模型
"""
from datetime import datetime
from app import db

class Task(db.Model):
    """任务模型"""
    __tablename__ = 'tasks'
    
    id = db.Column(db.Integer, primary_key=True)
    title = db.Column(db.String(255), nullable=False)
    description = db.Column(db.Text)
    status = db.Column(db.String(20), default='pending')  # pending, processing, completed, cancelled
    priority = db.Column(db.String(20), default='medium')  # low, medium, high, critical
    alert_id = db.Column(db.Integer, db.ForeignKey('alerts.id'))
    assigned_to = db.Column(db.Integer, db.ForeignKey('users.id'))
    location = db.Column(db.String(255))
    latitude = db.Column(db.Float)
    longitude = db.Column(db.Float)
    target_location = db.Column(db.String(255))
    target_latitude = db.Column(db.Float)
    target_longitude = db.Column(db.Float)
    evidence_count = db.Column(db.Integer, default=0)
    checklist_progress = db.Column(db.Integer, default=0)  # 0-100
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    started_at = db.Column(db.DateTime)
    completed_at = db.Column(db.DateTime)
    
    # 关系
    alert = db.relationship('Alert', backref='tasks')
    assigned_user = db.relationship('User', backref='tasks')
    
    def to_dict(self):
        """转换为字典"""
        return {
            'id': self.id,
            'title': self.title,
            'description': self.description,
            'status': self.status,
            'priority': self.priority,
            'alert_id': self.alert_id,
            'assigned_to': self.assigned_to,
            'location': self.location,
            'latitude': self.latitude,
            'longitude': self.longitude,
            'target_location': self.target_location,
            'target_latitude': self.target_latitude,
            'target_longitude': self.target_longitude,
            'evidence_count': self.evidence_count,
            'checklist_progress': self.checklist_progress,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
            'started_at': self.started_at.isoformat() if self.started_at else None,
            'completed_at': self.completed_at.isoformat() if self.completed_at else None
        }
    
    def __repr__(self):
        return f'<Task {self.title}>'

File: back-end/src/models/Task.js
import { DataTypes } from 'sequelize';
import { sequelize } from '../config/database.js';

const Task = sequelize.define('Task', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  title: {
    type: DataTypes.STRING(200),
    allowNull: false,
    validate: {
      len: {
        args: [1, 200],
        msg: '标题长度必须在1-200个字符之间'
      }
    }
  },
  description: {
    type: DataTypes.TEXT
  },
  type: {
    type: DataTypes.ENUM('quick-sampling', 'complete-sampling', 'inspection', 'monitoring'),
    allowNull: false
  },
  status: {
    type: DataTypes.ENUM('pending', 'in-progress', 'completed', 'cancelled'),
    defaultValue: 'pending'
  },
  priority: {
    type: DataTypes.ENUM('low', 'medium', 'high', 'urgent'),
    defaultValue: 'medium'
  },
  warning_id: {
    type: DataTypes.INTEGER,
    references: {
      model: 'alerts',
      key: 'id'
    }
  },
  alert_id: {
    type: DataTypes.INTEGER,
    references: {
      model: 'alerts',
      key: 'id'
    }
  },
  assigned_to: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: 'users',
      key: 'id'
    }
  },
  created_by: {
    type: DataTypes.INTEGER,
    references: {
      model: 'users',
      key: 'id'
    }
  },
  location: {
    type: DataTypes.STRING(255),
    allowNull: false
  },
  latitude: {
    type: DataTypes.DECIMAL(10, 8)
  },
  longitude: {
    type: DataTypes.DECIMAL(11, 8)
  },
  start_time: {
    type: DataTypes.DATE
  },
  end_time: {
    type: DataTypes.DATE
  },
  deadline: {
    type: DataTypes.DATE
  },
  progress: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
    validate: {
      min: 0,
      max: 100
    }
  },
  notes: {
    type: DataTypes.TEXT
  },
  completion_notes: {
    type: DataTypes.TEXT
  },
  joint_case_no: {
    type: DataTypes.STRING(64)
  },
  joint_departments: {
    type: DataTypes.JSON
  },
  collaboration_status: {
    type: DataTypes.ENUM('none', 'pending', 'in-progress', 'done'),
    defaultValue: 'none'
  },
  collaboration_notes: {
    type: DataTypes.TEXT
  }
}, {
  tableName: 'tasks',
  timestamps: true,
  underscored: true,
  indexes: [
    { fields: ['status'] },
    { fields: ['type'] },
    { fields: ['priority'] },
    { fields: ['assigned_to'] },
    { fields: ['warning_id'] },
    { fields: ['alert_id'] },
    { fields: ['joint_case_no'] },
    { fields: ['collaboration_status'] },
    { fields: ['created_at'] }
  ]
});

// 任务证据子表
const TaskEvidence = sequelize.define('TaskEvidence', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  task_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: 'tasks',
      key: 'id'
    }
  },
  type: {
    type: DataTypes.STRING(50)
  },
  url: {
    type: DataTypes.STRING(500)
  },
  description: {
    type: DataTypes.TEXT
  },
  uploaded_at: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW
  }
}, {
  tableName: 'task_evidence',
  timestamps: false,
  underscored: true
});

// 联合执法流转时间线
const TaskCollaborationFlow = sequelize.define('TaskCollaborationFlow', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  task_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: 'tasks',
      key: 'id'
    }
  },
  from_status: {
    type: DataTypes.ENUM('none', 'pending', 'in-progress', 'done')
  },
  to_status: {
    type: DataTypes.ENUM('none', 'pending', 'in-progress', 'done'),
    allowNull: false
  },
  action: {
    type: DataTypes.STRING(64),
    allowNull: false
  },
  operator_id: {
    type: DataTypes.INTEGER,
    references: {
      model: 'users',
      key: 'id'
    }
  },
  department: {
    type: DataTypes.STRING(128)
  },
  notes: {
    type: DataTypes.TEXT
  },
  payload: {
    type: DataTypes.JSON
  },
  created_at: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW
  }
}, {
  tableName: 'task_collaboration_flows',
  timestamps: false,
  underscored: true,
  indexes: [
    { fields: ['task_id', 'created_at'] },
    { fields: ['to_status', 'created_at'] }
  ]
});

// 任务检查清单子表
const TaskChecklist = sequelize.define('TaskChecklist', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  task_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: 'tasks',
      key: 'id'
    }
  },
  item: {
    type: DataTypes.STRING(255),
    allowNull: false
  },
  completed: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  },
  completed_at: {
    type: DataTypes.DATE
  }
}, {
  tableName: 'task_checklist',
  timestamps: true,
  underscored: true,
  createdAt: 'created_at',
  updatedAt: false
});

// 定义关联关系
Task.hasMany(TaskEvidence, { foreignKey: 'task_id', as: 'evidence' });
TaskEvidence.belongsTo(Task, { foreignKey: 'task_id' });

Task.hasMany(TaskChecklist, { foreignKey: 'task_id', as: 'checklist' });
TaskChecklist.belongsTo(Task, { foreignKey: 'task_id' });

Task.hasMany(TaskCollaborationFlow, { foreignKey: 'task_id', as: 'collaborationFlows' });
TaskCollaborationFlow.belongsTo(Task, { foreignKey: 'task_id' });

export { Task, TaskEvidence, TaskChecklist, TaskCollaborationFlow };
export default Task;
